### DISCUSSION TITLE
How to pass an ArrayBuffer object into server actions?

### DISCUSSION THREAD

**Discussion by DevinJohw** [vercel/next.js#66791] [Help]
### Summary

I want to use server actions to upload files, while I can't pass ArrayBuffer into server action, instead, Next.js throws an error like this:
<img width="1004" alt="image" src="https://github.com/vercel/next.js/assets/143488669/fcb61adb-039c-4a0b-8a32-7cd288266101">

Here's the code:
```javascript
function Upload() {
  const handleChange = async (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.currentTarget.files) {
      const file = e.currentTarget.files[0];
      await upload({
        arrayBuffer: await file.arrayBuffer(),
        fileName: file.name,
      });
    }
  };
  return (
    <div>
      <input type="file" onChange={handleChange} />
    </div>
  );
}
```

**Comment by icyJoseph** [comment]
You need to pass a `FormData` object to the `upload` function you have. That'll make it use the `multipart/form-data` content type, which would allow it to carry over the file. Otherwise, you are basically, asking it to serialize the file.

**Reply by DevinJohw** [reply]
Thanks! What if I want to do chunked upload, instead of uploading the entire file.

**Reply by icyJoseph** [reply]
I think you should be able to slice the chunks, and put them into the FormData: https://developer.mozilla.org/en-US/docs/Web/API/FormData/append

**Reply by DevinJohw** [reply]
Great! It seems I find a solution with the `formdata.append`, although It's not elegant enough I think. By the way, is it possible for server action to accept ArrayBuffer or Blob as argument? If possible, whether the Next.js will support this feature in the future? If impossible, could you explain me why this is impossible. Thanks in advance.

**Reply by icyJoseph** [reply]
You have to think that the server action is invoked, with some client side data, on the server, so the client data has to "travel" from the client to the server, and for that, it has to be a supported content type.
